plugins/BEdita/Core/src/Model/Table/TreesTable.php

```diff
@@ -99,9 +104,76 @@ public function validationDefault(Validator $validator)
     public function buildRules(RulesChecker $rules)
     {
         $rules->add($rules->existsIn(['object_id'], 'Objects'));
-        $rules->add($rules->existsIn(['parent_id'], 'ParentTrees'));
-        $rules->add($rules->existsIn(['root_id'], 'Objects'));
+        $rules->add($rules->existsIn(['root_id'], 'RootObjects'));
+        $rules->add($rules->existsIn(
+            ['parent_id'],
+            'ParentObjects',
+            ['allowNullableNulls' => true]
+        ));
+        $rules->add($rules->existsIn(
+            ['parent_node_id'],
+            'ParentNode',
+            ['allowNullableNulls' => true]
+        ));
+
+        $rules->add(
+            [$this, 'isParentValid'],
+            'isParentValid',
+            [
+                'errorField' => 'parent_id',
+                'message' => __d('bedita', 'parent_id must be null or corresponding to a folder'),
+            ]
+        );
 
         return $rules;
     }
+
+    /**
+     * Check that `parent_id` property of the entity corresponds to a folder
+     *
+     * @param EntityInterface $entity The entity to validate
+     * @return bool
+     */
+    public function isParentValid(EntityInterface $entity)
+    {
+        if ($entity->parent_id === null) {
+            return true;
+        }
+
+        $table = TableRegistry::get('Objects');
+        $isParentFolder = $table->find()
+            ->where([
+                $table->aliasField('id') => $entity->parent_id,
+            ])
+            ->innerJoinWith('ObjectTypes', function (Query $query) {
+                return $query->where(['ObjectTypes.name' => 'folders']);
+            })
+            ->count();
```

Comment:
Since we're using a strict condition on the primary key, it is fine to `COUNT(*)`. However, in these cases it is generally more efficient to have a query built this way:

```sql
SELECT 1 FROM mytable WHERE myfield = 'something' LIMIT 1;
```

This prevents SQL from loading more rows than needed, as we actually only want to know if a row that respects such conditions exists — whether it is one or a thousand is not relevant.

CakePHP's ORM has a `Table::exist($conditions)` method that could be used. In this case:

```php
return TableRegistry::get('Objects')->exists([
    'id' => $entity->parent_id,
    'object_type_id' => TableRegistry::get('ObjectTypes')->get('folders')->id,
]);
```

Reply:
Good point!